File: sagemaker-vllm-advisor/src/services/huggingface.ts
import type { AssessmentInput, Precision } from "../engine/types";

export interface HuggingFaceModelSummary {
  id: string;
  downloads: number;
  likes: number;
  gated: boolean | "auto" | "manual";
  pipelineTag?: string;
}

interface BackendModelMetadata {
  id: string;
  revision?: string | null;
  architecture: string;
  modelType: string;
  gated: boolean;
  parameterBillions?: number | null;
  layers?: number | null;
  hiddenSize?: number | null;
  attentionHeads?: number | null;
  kvHeads?: number | null;
  maxPositionEmbeddings?: number | null;
  precision?: Precision | null;
}

interface RawModelDetails {
  id: string;
  safetensors?: { total?: number };
}

interface RawModelConfig {
  architectures?: string[];
  model_type?: string;
  hidden_size?: number;
  n_embd?: number;
  num_hidden_layers?: number;
  n_layer?: number;
  num_attention_heads?: number;
  n_head?: number;
  num_key_value_heads?: number;
  max_position_embeddings?: number;
  n_positions?: number;
  torch_dtype?: string;
}

export interface LoadedModelMetadata {
  id: string;
  revision?: string;
  architecture: string;
  modelType: string;
  gated: boolean;
  patch: Partial<AssessmentInput>;
}

function precisionFromDtype(dtype?: string): Precision | undefined {
  const normalized = dtype?.toLowerCase();
  if (normalized === "bfloat16" || normalized === "bf16") return "BF16";
  if (normalized === "float16" || normalized === "fp16") return "FP16";
  if (normalized === "float8" || normalized === "fp8") return "FP8";
  return undefined;
}

export function metadataToAssessmentPatch(
  details: RawModelDetails,
  config: RawModelConfig
): Partial<AssessmentInput> {
  const attentionHeads = config.num_attention_heads ?? config.n_head;
  const contextLength =
    config.max_position_embeddings ?? config.n_positions;
  const precision = precisionFromDtype(config.torch_dtype);
  return {
    modelId: details.id,
    ...(details.safetensors?.total
      ? { parameterBillions: details.safetensors.total / 1_000_000_000 }
      : {}),
    ...(config.num_hidden_layers ?? config.n_layer
      ? { layers: (config.num_hidden_layers ?? config.n_layer)! }
      : {}),
    ...(config.hidden_size ?? config.n_embd
      ? { hiddenSize: (config.hidden_size ?? config.n_embd)! }
      : {}),
    ...(attentionHeads ? { attentionHeads } : {}),
    ...(config.num_key_value_heads
      ? { kvHeads: config.num_key_value_heads }
      : attentionHeads
        ? { kvHeads: attentionHeads }
        : {}),
    ...(precision ? { precision } : {}),
    ...(contextLength
      ? { maxModelLength: Math.min(contextLength, 8192) }
      : {})
  };
}

async function apiError(response: Response, fallback: string) {
  try {
    const body = (await response.json()) as { detail?: string };
    return body.detail ?? fallback;
  } catch {
    return fallback;
  }
}

export async function searchHuggingFaceModels(
  query: string,
  signal?: AbortSignal
): Promise<HuggingFaceModelSummary[]> {
  const trimmed = query.trim();
  if (trimmed.length < 2) return [];
  const response = await fetch(
    `/api/models/search?q=${encodeURIComponent(trimmed)}`,
    { signal }
  );
  if (!response.ok) {
    throw new Error(
      await apiError(response, `Hugging Face search failed (${response.status}).`)
    );
  }
  return response.json();
}

export async function loadHuggingFaceModel(
  modelId: string
): Promise<LoadedModelMetadata> {
  const trimmed = modelId.trim();
  if (!trimmed.includes("/")) {
    throw new Error("Enter a complete model ID such as Qwen/Qwen3-8B.");
  }
  const response = await fetch(
    `/api/models/resolve?model_id=${encodeURIComponent(trimmed)}`
  );
  if (!response.ok) {
    throw new Error(
      await apiError(
        response,
        `Model metadata could not be loaded (${response.status}).`
      )
    );
  }
  const model = (await response.json()) as BackendModelMetadata;
  const patch: Partial<AssessmentInput> = {
    modelId: model.id,
    ...(model.parameterBillions
      ? { parameterBillions: model.parameterBillions }
      : {}),
    ...(model.layers ? { layers: model.layers } : {}),
    ...(model.hiddenSize ? { hiddenSize: model.hiddenSize } : {}),
    ...(model.attentionHeads ? { attentionHeads: model.attentionHeads } : {}),
    ...(model.kvHeads ? { kvHeads: model.kvHeads } : {}),
    ...(model.precision ? { precision: model.precision } : {}),
    ...(model.maxPositionEmbeddings
      ? { maxModelLength: Math.min(model.maxPositionEmbeddings, 8192) }
      : {})
  };
  return {
    id: model.id,
    revision: model.revision ?? undefined,
    architecture: model.architecture,
    modelType: model.modelType,
    gated: model.gated,
    patch
  };
}

File: sagemaker-vllm-advisor/backend/providers/huggingface.py
from __future__ import annotations

import asyncio
from typing import Any
from urllib.parse import quote

import httpx

from backend.domain.models import (
    EvidenceRecord,
    ModelMetadata,
    ModelSearchResult,
    Precision,
)


def precision_from_dtype(dtype: str | None) -> Precision | None:
    normalized = (dtype or "").lower()
    if normalized in {"bfloat16", "bf16"}:
        return Precision.BF16
    if normalized in {"float16", "fp16"}:
        return Precision.FP16
    if normalized in {"float8", "fp8"}:
        return Precision.FP8
    return None


class HuggingFaceProvider:
    def __init__(self, token: str | None = None, timeout_seconds: float = 15):
        self._headers = {"Authorization": f"Bearer {token}"} if token else {}
        self._timeout = timeout_seconds

    async def search(self, query: str, limit: int = 8) -> list[ModelSearchResult]:
        params = {
            "search": query,
            "filter": "text-generation",
            "sort": "downloads",
            "direction": "-1",
            "limit": str(min(max(limit, 1), 25)),
        }
        async with httpx.AsyncClient(
            headers=self._headers, timeout=self._timeout
        ) as client:
            response = await client.get("https://huggingface.co/api/models", params=params)
            response.raise_for_status()
            return [
                ModelSearchResult(
                    id=item["id"],
                    downloads=item.get("downloads", 0),
                    likes=item.get("likes", 0),
                    gated=item.get("gated", False),
                    pipeline_tag=item.get("pipeline_tag"),
                )
                for item in response.json()
            ]

    async def resolve(self, model_id: str) -> ModelMetadata:
        encoded = "/".join(quote(part, safe="") for part in model_id.split("/"))
        async with httpx.AsyncClient(
            headers=self._headers,
            timeout=self._timeout,
            follow_redirects=True,
        ) as client:
            details_response, config_response = await asyncio.gather(
                client.get(f"https://huggingface.co/api/models/{encoded}"),
                client.get(
                    f"https://huggingface.co/{encoded}/resolve/main/config.json"
                ),
            )
        details_response.raise_for_status()
        details: dict[str, Any] = details_response.json()
        gated = bool(details.get("gated"))
        if config_response.status_code >= 400:
            if gated:
                raise PermissionError(
                    "This gated model requires a configured Hugging Face token."
                )
            config_response.raise_for_status()
        config: dict[str, Any] = config_response.json()

        total_parameters = (details.get("safetensors") or {}).get("total")
        revision = details.get("sha")
        architecture = (config.get("architectures") or ["Unknown architecture"])[0]
        attention_heads = config.get("num_attention_heads") or config.get("n_head")
        kv_heads = config.get("num_key_value_heads") or attention_heads
        context = config.get("max_position_embeddings") or config.get("n_positions")

        return ModelMetadata(
            id=details["id"],
            revision=revision,
            architecture=architecture,
            model_type=config.get("model_type", "unknown"),
            gated=gated,
            parameter_billions=(
                total_parameters / 1_000_000_000 if total_parameters else None
            ),
            layers=config.get("num_hidden_layers") or config.get("n_layer"),
            hidden_size=config.get("hidden_size") or config.get("n_embd"),
            attention_heads=attention_heads,
            kv_heads=kv_heads,
            max_position_embeddings=context,
            precision=precision_from_dtype(config.get("torch_dtype")),
            evidence=[
                EvidenceRecord(
                    name="model_revision",
                    value=revision,
                    evidence_type="huggingface_api",
                    source="Hugging Face Hub API",
                    source_reference=f"https://huggingface.co/{model_id}",
                ),
                EvidenceRecord(
                    name="model_config",
                    value=architecture,
                    evidence_type="huggingface_api",
                    source="Hugging Face config.json",
                    source_reference=(
                        f"https://huggingface.co/{model_id}/resolve/main/config.json"
                    ),
                ),
            ],
        )
